Add MachineTranslatedStateTextsHandler and translation domain services

SarahZLipeles opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #N/A
2. This PR does the following: Add translation domain services as well as MachineTranslatedStateTextsHandler and endpoint for the [Computer Aided Translation project](https://docs.google.com/document/d/1kJd-yLTzB9a2c3Nq7v9pzKfHwKHKGpkWfQ8B0YGf50U/edit?usp=sharing).

**BLOCKED BY**: PR #12515 

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Commit:
workspace transfer

Commit:
formatting

Commit:
move MachineTranslatedText to translation_domain.py

Commit:
import utils convert to hash

Commit:
Merge remote-tracking branch 'upstream/develop' into machine-translated-text

Commit:
add storage model

Commit:
add translation domain object and tests

Commit:
Merge remote-tracking branch 'upstream/develop' into machine-translated-text

Commit:
add codeowners

Commit:
fix typo

Commit:
change code owner to @vojtechjelinek

Commit:
add translation_validators

Commit:
sort imports
fix spacing

Commit:
fix origin_text type
Change max hash length to be more "in range"
Fix/clarify docstrings and comments
Add TODOs for tidying language code verification logic related to #12341

Commit:
fix spacing

Commit:
add MachineTranslatedTextModelAuditOneOffJob to job registry

Commit:
fix typo

Commit:
add hashed_origin_text field to MachineTranslatedTextModel
(allows lookup by origin_text)

Commit:
address pr comments:
change MachineTranslatedTextTests name to be consistent
clarify/clean-up comments
remove unnecessary backslashes from regex
fix spacing

Commit:
Merge remote-tracking branch 'upstream/develop' into machine-translated-text

Commit:
fix typo

Commit:
add translation fetchers

Commit:
add get_translation_for_text fetcher

Commit:
add translation_fetchers tests

Commit:
add translation services and tests

Commit:
fix spacing

Commit:
fix misnamed setup
sort codeowners lines

Commit:
Merge branch 'machine-translated-text' into get-machine-translated-text

Commit:
add MachineTranslatedTextHandler endpoint

Commit:
fix linter issues

SarahZLipeles commented:
@sagangwee @seanlip If you would like to begin reviewing this, look at commit e340cf9c254ae25125f72e50e708cad7fa24ecc2 ("add translation services and tests") and newer

Commit:
fix formating and docstring nits

Commit:
change _generate_id to be a static method

Commit:
add tests for get_translation_for_text
clarify docstrings

Commit:
consolidate terminology:
    replace "origin" with "source" in reference to the untranslated source text

Commit:
Merge branch 'machine-translated-text' into get-machine-translated-text

Commit:
add more description regarding key purpose

Commit:
replace ":" with "." as a divider character to conform to codebase practices

seanlip commented on .github/CODEOWNERS:
```diff
@@ -110,7 +110,9 @@
 /core/domain/opportunity*.py @sagangwee
+/core/domain/translation*.py @vojtechjelinek
 /core/domain/voiceover_services*.py @sagangwee
+/core/storage/translation/ @vojtechjelinek
```
Please don't include this line. All storage models should be owned by @BenHenning (see entry at the bottom of the codeowners file). It's fine to just delete this line.

SarahZLipeles replied:
Done.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,45 @@ def _content_in_review(self, state_name, content_id, suggestions):
+            self.values = {'translated_texts': translated_texts}
+            self.render_json(self.values)
+        else:
+            raise self.PageNotFoundException()
```
Should it be an empty dict? It seems to me that the cases where 1 vs 0 translations out of 20 were successfully provided should be handled the same way.

SarahZLipeles replied:
**Not** Fixed. I plan on doing some refactoring here for efficiency and debugging.

SarahZLipeles replied:
Refactor complete. I moved the get_machine_translation_for_content_id logic into MachineTranslatedStateTextHandler and updated the return values and tests accordingly.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,45 @@ def _content_in_review(self, state_name, content_id, suggestions):
+
+        # TODO(#12341): Tidy up this logic once we have a canonical list of
+        # language codes.
+        if not utils.is_supported_audio_language_code(
```
Is there a need to check first that target_language_code (and exp_id, state_name for that matter) are not None?

SarahZLipeles replied:
I believe not, `NoneType` should fail the type validations just like any other type. Is this worth adding tests for?

seanlip replied:
Yup, worth adding tests. Thanks!

SarahZLipeles replied:
I added a variety of tests. PTAL.

seanlip commented on core/domain/translation_domain.py:
```diff
@@ -0,0 +1,118 @@
+            raise utils.ValidationError(
+                'Invalid target language code: %s' % self.target_language_code)
+
+        if self.source_language_code is self.target_language_code:
```
Probably should use `==` here, since you're comparing equality of two strings which isn't an "is-a" thing.

SarahZLipeles replied:
Fixed. I'm surprised `test_validate_with_same_source_target_language_codes_raises` didn't fail.

seanlip commented on core/domain/translation_domain_test.py:
```diff
@@ -0,0 +1,110 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for translate domain objects."""
```
Tests for domain objects related to translations.

SarahZLipeles replied:
Fixed.

seanlip commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,102 @@
+            translation language. Must be different from source_language_code.
+
+    Returns:
+        str. The translated text.
```
Only one value is returned, so you should write: `str|None. The translated text, or None if no translation is found.`

SarahZLipeles replied:
Fixed.

seanlip commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,102 @@
+
+    if state_name not in state_names_to_content_id_mapping:
+        return None
+    state = state_names_to_content_id_mapping[state_name]
```
This object is not a state, please name it properly according to what it actually represents.

SarahZLipeles replied:
Fixed.

seanlip commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,102 @@
+        ValueError. Invalid language code.
+
+    Returns:
+        str. The translated text.
```
Ditto, same comment as above.

SarahZLipeles replied:
Fixed.

seanlip commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,102 @@
+    translation = translation_fetchers.get_translation_for_text(
+        source_language_code,
+        target_language_code,
+        source_text.strip()
```
Just a question. How are you planning to handle the case where the source text has e.g. images -- does Cloud Translate preserve those tags appropriately?

SarahZLipeles replied:
Cloud Translate defaults to assuming content is in html format and ignores tags, only translating text between tags. I'll keep an eye out for special cases though.
edit: It doesn't strip the tags, it just leaves them as is, for example, I tried running this using my GCP free tier:
```py
>>> print(translate_text('<p>hello world</p><img src="world.png">', 'en', 'es'))
<p>Hola Mundo</p><img src="world.png">
```

seanlip replied:
Ah, I see -- thanks! Maybe you should leave a comment for that somewhere, it's useful to know. Perhaps in the platform layer.

SarahZLipeles replied:
Done.

seanlip commented on core/domain/translation_services_test.py:
```diff
@@ -0,0 +1,131 @@
+        )
+        self.assertEqual(translated_text, 'Por favor continua.')
+
+    def test_get_machine_translation_for_content_id_with_invalid_exploration_id_returns_none(self): # pylint: disable=line-too-long
```
`test_get_machine_translation_is_none_if_exp_id_is_invalid`

I think you can do similar names below. Try to avoid adding pylint pragmas.

SarahZLipeles replied:
Done. Could this cause confusion since I'm testing `get_machine_translation_for_content_id` and `get_machine_translation` is a separate method?

seanlip replied:
Oh, good point. Yes, I think it might. Maybe test_get_content_id_translation or test_get_translation_for_content_id perhaps -- the method name is really long and it's probably a good idea to find a reasonable way to abbreviate it.

SarahZLipeles replied:
This is no longer relevant since `get_machine_translation_for_content_id` is removed.

seanlip commented on core/domain/translation_validators.py:
```diff
@@ -0,0 +1,51 @@
+        language_codes_regex = '(%s)' % '|'.join(
+            cloud_translate_services.LANGUAGE_CODE_ALLOWLIST)
+        # Valid id: [source_language_code]:[target_language_code]:[hashed_text].
+        regex_string = '^%s:%s:.+$' % (
```
There was some discussion in the other PR about whether the delimiter should be a colon or a dot; what was the eventual decision?

SarahZLipeles replied:
We ended up going with `.` as of a recent commit.

Commit:
Merge branch 'machine-translated-text' into get-machine-translated-text

Commit:
fix comments and variable names
remove incorrect CODEOWNERS line
fix incorrect string comparison

Commit:
Merge remote-tracking branch 'upstream/develop' into get-machine-translated-text

Commit:
rename MachineTranslatedStateTextsHandler for consistency with design doc

Commit:
refactor MachineTranslatedStateTextHandlerTests
move get_machine_translation_for_content_id logic into MachineTranslatedStateTextHandler

Commit:
add comment clarifying html handling

SarahZLipeles commented:
@seanlip I addressed your comments, PTAL.

SarahZLipeles commented:
@sagangwee @brianrodri @DubeySandeep @nithusha21 @U8NWXD @vojtechjelinek PTAL. I'm aiming to have this merged by the end of today if possible. (This is the last day of my rotation with Oppia)

nithusha21 reviewed (approved):
LGTM for codeowner files!

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
 
 
+class MachineTranslatedStateTextsHandler(base.BaseHandler):
+    """Provide a machine translation of exploration content."""
```
Nit: Use third-person, i.e. "Provides..."

SarahZLipeles replied:
Done.

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
+        translation of form:
+            dict('translated_texts', dict(str, str|None))
+        If no translation is found for a given content id, that id is mapped to
+        None.
```
Can we add the Args, Returns, and Raises here?

https://github.com/oppia/oppia/issues/11776

SarahZLipeles replied:
Hm, there are no args and the method technically doesn't return anything. Would "Request" and "Response" make more sense?

SarahZLipeles replied:
Alternatively "Params" and "Data Response" to mirror the language in the design doc.

SarahZLipeles replied:
I went with the latter, PTAL.

sagangwee commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -487,6 +488,149 @@ def test_handler_does_not_return_in_review_content(self):
+        }
+        self.assertEqual(output, expected_output)
+
+    def test_handler_with_invalid_content_ids_format_returns_raises_exception(
```
Remove "returns"?

SarahZLipeles replied:
Fixed.

sagangwee commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -487,6 +488,149 @@ def test_handler_does_not_return_in_review_content(self):
+
+        self.publish_exploration(self.owner_id, exp.id)
+
+    def test_handler_with_invalid_language_code_raises_exception(self):
```
Can we test the exception string to make sure the correct exception is raised? Here and below.

SarahZLipeles replied:
Added tests.

sagangwee commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,67 @@
+    Returns:
+        str|None. The translated text or None if no translation is found.
+    """
+
```
Remove this line?

SarahZLipeles replied:
Fixed.

sagangwee commented on core/domain/translation_fetchers.py:
```diff
@@ -43,3 +43,30 @@ def get_translation_from_model(translation_model):
+        MachineTranslatedText|None. The MachineTranslatedText
+        if a translation exists or None if no translation is found.
+    """
+
```
Remove this line?

SarahZLipeles replied:
Fixed.

sagangwee commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,67 @@
+def get_machine_translation(
+        source_language_code, target_language_code, source_text):
+    """Gets a machine translation of the source text for the given source and
+    target languages.
```
Can we note that this method will also translate via the Cloud Translate API and populate a new MachineTranslation if the translation is not found? And update the method name accordingly?

SarahZLipeles replied:
Added.

vojtechjelinek reviewed (approved):
LGTM!

brianrodri reviewed (approved):
LGTM for codeowner files (1)

sagangwee commented on core/domain/translation_fetchers_test.py:
```diff
@@ -37,12 +38,22 @@ def test_get_translation_from_model(self):
+            'en', 'es', 'untranslated_text')
+        self.assertIsNone(translation)
+
+    def test_get_translation_for_text_with_existing_translation(
```
Can you add the expected outcome in the test name?

SarahZLipeles replied:
Added.

sagangwee commented on core/domain/translation_fetchers.py:
```diff
@@ -43,3 +43,30 @@ def get_translation_from_model(translation_model):
         translation_model.translated_text)
+
+
+def get_translation_for_text(
```
`get_machine_translation` might be clearer here.

seanlip replied:
get_machine_translated_text perhaps? Should match the domain object type.

Having said that ... I just realized that the object name is quite unclear for this reason. It's a "text", so I was expecting this function to return a string and was quite surprised to find that it didn't. I wonder if it's worth renaming the model and corresponding domain object to MachineTranslation / AutogeneratedTranslation, which seems more like a "thing" than a string? Sorry for not catching this before.

SarahZLipeles replied:
I changed this to `get_machine_translated_text`. I have to go for a bit though, I'll take another look at changing the model names later tonight.

SarahZLipeles replied:
@seanlip I renamed these, PTAL.

sagangwee commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -487,6 +488,149 @@ def test_handler_does_not_return_in_review_content(self):
+                'target_language_code': 'es'
+            },
+            expected_status_int=200
+        )
```
Add a new line after this one?

SarahZLipeles replied:
Added.

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
+        if not state_names_to_content_id_mapping.has_key(state_name):
+            raise self.PageNotFoundException(
+                'Invalid state_name: %s' % state_name)
+        content_id_mapping = state_names_to_content_id_mapping[state_name]
```
Might be clearer if you also include what the content ID maps to in the variable name.

SarahZLipeles replied:
Added.

sagangwee reviewed (changes requested):
@TheoLipeles Took a pass, thanks! Mostly just nits.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
+    def get(self):
+        """Handles GET requests. Responds with a mapping from content id to
+        translation of form:
+            dict('translated_texts', dict(str, str|None))
```
Nit: add a blank line above this, and a blank line below this.

SarahZLipeles replied:
Added.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
+            raise self.InvalidInputException('Missing state_name')
+
+        content_ids_string = self.request.get('content_ids')
+        content_ids = None
```
Is this line needed? If it is, should it default to `[]`?

SarahZLipeles replied:
This line is only here to scope content_ids outside of the try catch. `[]` works a lot better.
Fixed.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
+            raise self.PageNotFoundException('Invalid exp_id: %s' % exp_id)
+        state_names_to_content_id_mapping = exp.get_translatable_text(
+            target_language_code)
+        if not state_names_to_content_id_mapping.has_key(state_name):
```
Here and elsewhere: I think has_key is deprecated in Python 3. Just use "in".

Ditto below/elsewhere.

SarahZLipeles replied:
Fixed.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -260,6 +263,73 @@ def _content_in_review(self, state_name, content_id, suggestions):
+                    translation_services.get_machine_translation(
+                        exp.language_code, target_language_code, source_text)
+                )
+            else:
```
Should this case be considered an error? It would correspond to one of the input content IDs being invalid, right?

SarahZLipeles replied:
I modified this to be the catch-all case for the translation call. Rather than raising an error at the first problematic ID, I think it's better to return the translations retrieved and mark unsuccessful translations None.
This is specifically considering the fact that returning a 404 does not allow custom error messages, which could cause issues with debugging which content id caused the problem.

seanlip commented on main.py:
```diff
@@ -340,6 +340,9 @@ def ui_access_wrapper(self, *args, **kwargs):
     get_redirect_route(
         r'%s/story/<story_url_fragment>' % feconf.TOPIC_VIEWER_URL_PREFIX,
         story_viewer.StoryPage),
+    get_redirect_route(
```
This seems to be placed somewhat randomly in the list. Put this near the contributor dashboard URLs, and follow the existing convention of having the names defined as constants in feconf. Follow existing naming patterns/styles.

SarahZLipeles replied:
Fixed, but I'm a little confused what the convention is, some urls seem to use feconf constants while others do not.

seanlip replied:
Yeah, that's fair -- I forgot we hadn't fixed this universally yet. What you have for now is fine for now, I think -- thanks!

Commit:
clarify MachineTranslatedStateTextsHandlerTests error messages

Commit:
rename get_translation_for_text to get_machine_translation
add return value to test name

Commit:
fix spacing and naming

Commit:
fix linter errors

Commit:
rename translation_services.get_machine_translation to translation_services.get_and_cache_machine_translation
clarify docstrings.
fix spacing.

Commit:
rename translation_fetchers.get_machine_translation to get_machine_translated_text

SarahZLipeles commented:
@sagangwee I addressed your comments, PTAL.

DubeySandeep reviewed (approved):
I only have only one comment in `main.py` which @seanlip has already mentioned! (Approving!)

Commit:
change MachineTranslatedText to MachineTranslation
refactor content_id error response for MachineTranslationStateTextsHandler
fix formatting nits

SarahZLipeles commented:
@seanlip I addressed your comments, PTAL.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -263,18 +264,47 @@ def _content_in_review(self, state_name, content_id, suggestions):
         None.
+
+        Params:
+            exp_id: str. The ID of exploration being translated.
```
Nit: add "the" before "exploration"

SarahZLipeles replied:
Fixed.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -263,18 +264,47 @@ def _content_in_review(self, state_name, content_id, suggestions):
+        Data Response:
+
+            dict(
+                'translated_texts', dict(str, str|None),
```
After 'translated_texts' thre should be a colon, not a comma.

SarahZLipeles replied:
Fixed.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -263,18 +264,47 @@ def _content_in_review(self, state_name, content_id, suggestions):
+                from content ID to the translated text. If an error occured
+                during retrieval of some content translations, but not others,
+                failed translations are mapped to None and the message is
+                returned in the 'errors' dict as a mapping from content id to
```
I don't think you need to do this. Just return translated_texts with failed translations mapped to None, and do a logging.error() if there are any errors. The latter will be picked up in StackDriver and will be investigated. The user can't do anything about this anyway.

SarahZLipeles replied:
Done.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -308,25 +338,32 @@ def get(self):
-                source_text = content_id_mapping[content_id]
+            try:
+                if content_id not in content_id_to_text_mapping:
+                    raise LookupError('Invalid content_id: %s' % content_id)
```
Anything raised at this level should be self.InvalidInputException or self.PageNotFoundException etc., following the usual conventions for 400, 404, 401. In this case I think it's a 400.

SarahZLipeles replied:
Wouldn't returning an exception here prevent a partially successful translation from returning the successful translations?

SarahZLipeles replied:
Fixed. See comments below.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -308,25 +338,32 @@ def get(self):
-            else:
+            except Exception as e:
                 translated_texts[content_id] = None
+                errors[content_id] = python_utils.UNICODE(e)
```
Don't think you need this, see above.

SarahZLipeles replied:
Removed.

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -308,25 +338,32 @@ def get(self):
         for content_id in content_ids:
-            if content_id_mapping.has_key(content_id):
-                source_text = content_id_mapping[content_id]
+            try:
```
The part in the "try" part of a try/except should be as small as possible. Which part exactly are you thinking might fail here?

Note that "extra" content IDs are just a regular 400 error. It means that the client has sent bad input.

SarahZLipeles replied:
I was thinking that `translation_services.get_and_cache_machine_translation` might error, but I forgot I set it up to return None for an unsuccessful translation. I'll also return None for an "extra" content ID.

SarahZLipeles replied:
I removed the try catch, this now returns None for failures to translate specific content_ids with no error info sent to the front end.

seanlip commented on core/domain/translation_fetchers_test.py:
```diff
@@ -32,28 +32,28 @@ class TranslationFetchersTests(test_utils.GenericTestBase):
+        translation_models.MachineTranslationModel.create(
             'en', 'es', 'hello world', 'hola mundo')
-        translation = translation_fetchers.get_translation_for_text(
+        translation = translation_fetchers.get_machine_translated_text(
```
Should this be get_machine_translation() (is it fetching the translation object or the text)?

SarahZLipeles replied:
Fixed, thanks for catching this!

Commit:
fix leftover renaming issue
remove duplicate error handling
fix coment format

SarahZLipeles commented:
@seanlip I addressed your comments, PTAL.

sagangwee commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -487,6 +488,189 @@ def test_handler_does_not_return_in_review_content(self):
+                'content_ids': '["invalid_content_id", "content"]',
+                'target_language_code': 'es'
+            }, expected_status_int=200
+        )
```
Add a new line here.

SarahZLipeles replied:
Done.

sagangwee commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,67 @@
+        source_language_code, target_language_code, source_text):
+    """Gets a machine translation of the source text for the given source and
+    target languages. If no translation exists in the datastore for the given
+    input, calls cloud_translate_services.
```
"and populates the translation..." or something to that effect.

SarahZLipeles replied:
Done.

sagangwee reviewed (approved):
@TheoLipeles LGTM, just a couple minor comments. Thanks!

Commit:
clarify docstring
fix spacing

seanlip commented on core/domain/translation_services.py:
```diff
@@ -0,0 +1,67 @@
+        ValueError. Invalid language code.
+
+    Returns:
+        str|None. The translated text or None if no translation is found.
```
I don't understand this -- at which point do we return None? If it's from cloud_translate_services.translate_text() then why are we creating a storage model with a None translation?

SarahZLipeles replied:
Fixed.

seanlip reviewed (changes requested):
Thanks @TheoLipeles, just one comment left.

Commit:
remove error raising from get_and_cache_machine_translation

SarahZLipeles commented:
@seanlip I fixed the problem you mentioned, PTAL.

seanlip commented on core/domain/translation_services.py:
```diff
@@ -56,13 +55,19 @@ def get_and_cache_machine_translation(
+    try:
+        translated_text = cloud_translate_services.translate_text(
+            source_text, source_language_code, target_language_code)
+    except ValueError as e: # Language code not allowlisted
```
Please put the comment above this line (not at the side), have it end with a period, and provide a bit more detail as a complete sentence. Also explain why we merely log the error instead of letting it raise, so that future developers have context on why the decision was made.

SarahZLipeles replied:
Done.

seanlip commented on core/domain/translation_services.py:
```diff
@@ -56,13 +55,19 @@ def get_and_cache_machine_translation(
+    except ValueError as e: # Language code not allowlisted
+        logging.error(e)
+
+    if translated_text is not None:
```
There should be a test that verifies that no model gets created for "the translated_text is not None" case. Always try to test any new behaviour and the different branches; line coverage alone is insufficient.

SarahZLipeles replied:
Thanks for catching this! Done.

Commit:
clarify comment
improve branch test coverage

SarahZLipeles commented:
@seanlip PTAL.

Commit:
clarify comments

BenHenning reviewed (approved):
Thanks @TheoLipeles. Approved for codeowners.

Apologies for the delayed review--not sure why GitHub didn't auto-add me as a reviewer.
